lib/kitchen/driver/gce.rb

```diff
@@ -16,154 +17,507 @@
       end
 
-      private
+      def validate!
+        raise "Either zone or region must be specified" unless config[:zone] || config[:region]
+        raise "'any' is no longer a valid region" if config[:region] == "any"
+        raise "Zone #{config[:zone]} is not a valid zone" if config[:zone] && !valid_zone?
+        raise "Region #{config[:region]} is not a valid region" if config[:region] && !valid_region?
+        raise "Machine type #{config[:machine_type]} is not valid" unless valid_machine_type?
+        raise "Disk type #{config[:disk_type]} is not valid" unless valid_disk_type?
+        raise "Network #{config[:network]} is not valid" unless valid_network?
+        raise "Email address of GCE user is not set" if winrm_transport? && config[:email].nil?
+
+        warn("Both zone and region specified - region will be ignored.") if config[:zone] && config[:region]
+      end
 
       def connection
-        options = {
-          provider: 'google',
-          google_client_email: config[:google_client_email],
-          google_project: config[:google_project]
-        }
+        return @connection unless @connection.nil?
 
-        [
-          :google_key_location,
-          :google_json_key_location
-        ].each do |k|
-          options[k] = config[k] unless config[k].nil?
+        @connection = Google::Apis::ComputeV1::ComputeService.new
+        @connection.authorization = authorization
+        @connection.client_options = Google::Apis::ClientOptions.new.tap do |opts|
+          opts.application_name    = "kitchen-google"
+          opts.application_version = Kitchen::Driver::GCE_VERSION
         end
 
-        Fog::Compute.new(options)
+        @connection
       end
 
-      def create_disk
-        disk = connection.disks.create(
-          name: config[:inst_name],
-          size_gb: config[:disk_size],
-          zone_name: config[:zone_name],
-          source_image: config[:image_name]
+      def authorization
+        @authorization ||= Google::Auth.get_application_default(
+          [
+            "https://www.googleapis.com/auth/cloud-platform",
+            "https://www.googleapis.com/auth/compute",
+          ]
         )
+      end
+
+      def winrm_transport?
+        instance.transport.name.downcase == "winrm"
+      end
+
+      def update_windows_password(server_name)
+        return unless winrm_transport?
+
+        username = instance.transport[:username]
+
+        info("Resetting the Windows password for user #{username} on #{server_name}...")
+
+        state[:password] = GoogleComputeWindowsPassword.new(
+          project:       project,
+          zone:          zone,
+          instance_name: server_name,
+          email:         config[:email],
+          username:      username
+        ).new_password
+
+        info("Password reset complete on #{server_name} complete.")
+      end
+
+      def check_api_call(&block)
+        block.call
+      rescue Google::Apis::ClientError
```

Comment:
@Temikus - I added a project validity check in 0d18b12.

Also, I opted not to add any error logging to `#check_api_call` because, in some cases, the driver uses that to check to see if something exists (such as an image) and it not existing is "okay." However, that exception message may still be useful in debugging purposes, so I opted to log it with debug severity instead. It will be there if the user runs test kitchen with `-l debug`. I hope that's a fair compromise!